voice implementation

dislocated-su opened the pull request:
Adds a lot of new dependencies for discordjs/voice.

- Adds music bot features to the bot

- Adds path aliases to the project (like `@voice/utils`)

- Refactors some utils

Old voice code still present but will be removed soon.

dislocated-su committed:
(does not work) attempted to implement @discordjs/voice

dislocated-su commented:
Would love some help on this, included some information about the problem in commands/slash/playnoise.ts

FYI, the changes to the environment are as follows:
+ @discordjs/opus
+ @discordjs/voice
+ libsodium-wrappers

ffmpeg is added via the dockerfile

if you figure out how to do this without @discordjs/voice that is also fine

dislocated-su committed:
voice uses streams instead of files

dislocated-su committed:
add subscription and track

dislocated-su committed:
disable old play command

dislocated-su committed:
(WIP) add /play and /skip

dislocated-su committed:
cleanup and polish play

dislocated-su committed:
chore: combine dockerfile apk commands

dislocated-su commented:
rebased branch onto master

dislocated-su committed:
chore: bump discordjs

dislocated-su committed:
chore: path aliases

dislocated-su committed:
chore: rename old commands

dislocated-su committed:
feat: additional voice commands

added leave, pause, resume, queue

dislocated-su committed:
fix: include m.youtube.com in regex

dislocated-su committed:
chore: remove unnecessary dependencies

TimMcGilly commented on src/commands/slash/voice/play.ts:
```diff
@@ -0,0 +1,137 @@
+    }
+
+
+    if (!subscription) {
+        if (e.interaction.member instanceof GuildMember && e.interaction.member.voice.channel) {
+            const channel = e.interaction.member.voice.channel;
+            subscription = new MusicSubscription(
+                joinVoiceChannel({
+                    channelId: channel.id,
+                    guildId: channel.guild.id,
+                    adapterCreator: createDiscordJSAdapter(channel as VoiceChannel),
+                }),
+            );
+            subscription.voiceConnection.on('error', console.warn);
+            subscriptions.set(e.interaction.guildId, subscription);
+        }
+    }
+    
+    // If there is no subscription, tell the user they need to join a channel.
+    if (!subscription) {
+        await e.interaction.followUp('Join a voice channel and then try that again!');
+        return;
+    }
```
```suggestion
    if (!subscription) {
        if (e.interaction.member instanceof GuildMember && e.interaction.member.voice.channel) {
            const channel = e.interaction.member.voice.channel;
            subscription = new MusicSubscription(
                joinVoiceChannel({
                    channelId: channel.id,
                    guildId: channel.guild.id,
                    adapterCreator: createDiscordJSAdapter(channel as VoiceChannel),
                }),
            );
            subscription.voiceConnection.on('error', console.warn);
            subscriptions.set(e.interaction.guildId, subscription);
        }
        else {
            await e.interaction.followUp('Join a voice channel and then try that again!');
            return;
        }
    }
```

dislocated-su committed:
chore: implement changes from review

Co-authored-by: Tim McGilly <11339148+TimMcGilly@users.noreply.github.com>

dislocated-su committed:
chore: add strict null checks

dislocated-su committed:
big commit: refactor and add channel check

change util locations:
* interfaces moved to shared/interactions
* removed getAttachment, moved getStream to voice/utils

dislocated-su committed:
fix: actually call the function in play.ts

dislocated-su committed:
fix: remove dangling void

dislocated-su committed:
fix: remove another dangling void

dislocated-su commented:
Not merging as I'm working on #19, as well as doing a small refactor and implementing feedback from user tests.

tomkettle1 commented on src/commands/slash/voice/play.ts:
```diff
@@ -0,0 +1,139 @@
+
+export const data: ApplicationCommandData = {
+    name: 'play',
+    description: 'Plays a song from url or from an attachment. To send an attachment, reply within 10s.',
```
```suggestion
    description: 'Plays a song from url or from an attachment. To send an attachment, reply within 10 seconds.',
```

dislocated-su committed:
feat(voice): add playlist support
feat(voice): split /play into two separate commands
fix(voice): /clear causes further queue flood on concurrent requests

dislocated-su commented:
Ok this PR now closes #18 and #19

TimMcGilly commented on package.json:
```diff
@@ -4,16 +4,26 @@
-    "discord.js": "dev",
+    "@discordjs/voice": "^0.5.4",
+    "@zerollup/ts-transform-paths": "^1.7.18",
+    "discord.js": "^13.0.0-dev.c8ca7bf.1627041776",
```
is this change to specific version correct?

dislocated-su replied:
Npm keeps changing it automatically whenever I bump the version. I'm not sure what we should go with.

dislocated-su replied:
changed it back to dev (for now at least)

TimMcGilly commented on src/shared/interactions/utils.ts:
```diff
@@ -0,0 +1,16 @@
+ * @returns True if user is in correct channel, false otherwise
+ */
+export function userInCorrectChannel(e: cmdEvent, subscription: MusicSubscription) : Boolean {
+    return !(subscription.voiceConnection.state.status === VoiceConnectionStatus.Ready
+        && subscription.voiceConnection.joinConfig.channelId !== (e.interaction.member as GuildMember).voice.channelId)
+}
```
Is this correct? Like in case of VoiceConnectionStatus.Disconnected wouldn't this return True?

dislocated-su replied:
You're right, there's no point double checking the connection is ready at that point.

TimMcGilly commented on src/shared/voice/utils.ts:
```diff
@@ -0,0 +1,147 @@
+	// Make sure the connection is ready before processing the user's request
+	try {
+		await entersState(subscription.voiceConnection, VoiceConnectionStatus.Ready, 15e3);
+		// Unlock the queue so new requests go through
+		subscription.queueLock = false;
```
Why do we need a unlock here? Where would it have been locked?

dislocated-su replied:
In a /clear. /clear would call `stop()` This will unlock the queue when someone actually wants to add a track.

TimMcGilly commented on src/shared/voice/subscription.ts:
```diff
@@ -0,0 +1,152 @@
+	public stop() {
+		this.queueLock = true;
+		this.queue = [];
+		this.audioPlayer.stop(true);
```
Feel like should unlock queueLock here

dislocated-su replied:
When using playlists, if the queue is unlocked right after a clear, and the playlist is still loading, it will load the rest into the queue. That's why it's not unlocked immediately.

dislocated-su replied:
It now unlocks queuelock here.

TimMcGilly commented on src/commands/slash/voice/play/url.ts:
```diff
@@ -0,0 +1,106 @@
+    // Regex: https://regexr.com/62j9j
+    const isYtVid = urlInput.match(/((http(s)?:)(\/\/))?(www\.)?((m.)?youtube.com|youtu.be)\/(watch)?(\?v=)?(\S+)?/) != null;
+
+    // This is defined only for YT videos
```
What is this comment "     // This is defined only for YT videos " refering to?

dislocated-su replied:
This comment is a leftover, will remove.

dislocated-su committed:
fix(voice): implement changes from review

dislocated-su committed:
fix(subscription): fix typo

dislocated-su committed:
chore: default slash commands to guild id of the message

dislocated-su committed:
fix: fix slash command registration race condition

tomkettle1 reviewed (approved):
they made me do it

dislocated-su committed:
chore(deps): bump discord.js and change req to dev label

dislocated-su committed:
fix(voice): update adapter to work with latest @discordjs/voice

dislocated-su committed:
fix: the missing perms error will no longer be uncaught

dislocated-su committed:
chore: remove old play command

dislocated-su committed:
chore: clean up voice utils
